### DISCUSSION TITLE
Does fluvio offer ordering gurantee? refer the below for NAT vs Kafka

### DISCUSSION THREAD

**Discussion by kharabela** [fluvio-community/fluvio#3889] [Q&A] [answered]
Ordering Assurance

NATS
Messages from a single publisher are delivered in the order they’re created to the subscriber. For multiple publishers, messages are out of order for different subscribers, and there’s no guarantee that messages will be delivered in the order they were published.
While out-of-order message delivery can be limiting in some use cases that require strict message ordering, it offers high performance and scalability benefits. For instance, you need message ordering to build control systems in a factory setting, but whether temperature readings from different sensors are delivered in a particular order is less critical.

For temperature, you need the latest measurement to be delivered on time, but it’s not important that an old reading is received last. By including timestamps in the message so that older timestamps can just be discarded, you can mitigate the ordering challenge in cases where it isn’t needed.

2.Kafka
Kafka is highly recommended for applications where maintaining order is a critical requirement, for example, in log aggregation, complex event processing (CEP), and staged event-driven architecture (SEDA).

In these use cases, Kafka provides an ordering guarantee for messages within a partition, ensuring that messages are delivered to consumers in the same order as they were produced. This is achieved by maintaining an ordered sequence of records within a partition, ensuring the exact order of messages is preserved.

**Comment by drc-infinyon** [comment, marked as answer]
Hi @kharabela,

Thanks for the question. I responded to the other question you asked. Sharing the answer here as well:

I created a similar image comparing NATS, Kafka, and Fluvio. Going to use it to update our documentation.
![InfinyOn Comparison](https://github.com/infinyon/fluvio/assets/122817612/b2d9bcea-51d5-4bbd-b289-2ac376fd2446)

In terms of Delivery Guarantees Fluvio supports at least once, at most once, and we support indempotence using deduplication which delivers the same output as exactly once - https://www.fluvio.io/docs/concepts/delivery-semantics/

Fluvio also supports ordering guarantees, persistence, replay-ability - https://www.fluvio.io/docs/concepts/data-consistency/

Throughput assumes a number of partitions and Fluvio can scale up to Kafka level throughputs while consuming a fraction of compute and memory.

On top of that the latest Stateful Services Development Kit enables stateful stream processing with offsets, states, and window based aggregations.

Hope this answers your question. Please let me know if you have any follow up.
